### PR TITLE
Implement Diff matchers

### PR DESCRIPTION
**Pull request by krzysztofzablocki**
When working on tests I often find myself having to scan a lot of logs just to see whats different, this is annoying. 

I've implemented diffing for tests.

From this:
before

to this:

after

### CODE REVIEW

File: Sourcery/CodeGenerated/Diffable.generated.swift [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        var results = DiffableResult()
+        guard let rhs = object as? Enum else {
+          results.append("Incorrect type, expected Enum, received: \(type(of: object))")
+          return results
+        }
+        if self.cases != rhs.cases { results.append("Incorrect cases, expected \(self.cases), received: \(rhs.cases)") }
+        if self.rawType != rhs.rawType { results.append("Incorrect rawType, expected \(self.rawType), received: \(rhs.rawType)") }
+
+        results.append(contentsOf: super.diffAgainst(object))
+        return results
```
Comment (ilyapuchka) [inline Sourcery/CodeGenerated/Diffable.generated.swift]: I think we need to call `super` here

File: Sourcery/Models/Diffable.swift [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+    /// Returns `DiffableResult` for the given objects.
+    ///
+    /// - Parameter object: Object to diff against.
+    /// - Returns: Diffable results.
+    func diffAgainst(_ object: Any?) -> DiffableResult
+}
+
+/// Phantom protocol for code generation
+protocol AutoDiffable {}
```
Comment (ilyapuchka) [inline Sourcery/Models/Diffable.swift]: How do you plan to use that? Looks like just having `Diffable` will be enough, no?

**Comment by ilyapuchka** [comment]
Looks very cool and useful indeed! What else are you planning to do here to make it not `wip`?

File: Sourcery/Models/Diffable.swift (hunk 2 above)
Comment (krzysztofzablocki) [inline Sourcery/Models/Diffable.swift]: In our case, but I'd like to write template a bit more generic than that as we could put it as example then

File: Sourcery/CodeGenerated/Diffable.generated.swift (hunk 1 above)
Comment (krzysztofzablocki) [inline Sourcery/CodeGenerated/Diffable.generated.swift]: yeah I kind of needs to add `super` access to `Type` so we can do things like `{% if type.super.implements.Diffable %}`

**Comment by krzysztofzablocki** [comment]
access to supertype and maybe allow recursive diff for properties that are composed

**Comment by krzysztofzablocki** [comment]
@ilyapuchka ready for review :)

**Pull request merged**